File: modules/ml-kit-digital-ink/ios/MlKitDigitalInkModule.swift
import ExpoModulesCore
import MLKit

public class MlKitDigitalInkModule: Module {
  private let modelManager = ModelManager.modelManager()
  private var modelCache: [String: DigitalInkRecognitionModel] = [:]
  private var recognizerCache: [String: DigitalInkRecognizer] = [:]

  public func definition() -> ModuleDefinition {
    Name("MlKitDigitalInk")

    AsyncFunction("isNativeRecognitionAvailableAsync") {
      true
    }

    AsyncFunction("prepareModelAsync") { (languageTag: String, promise: Promise) in
      do {
        let model = try self.getOrCreateModel(languageTag: languageTag)
        self.ensureModelReady(model: model, languageTag: languageTag, promise: promise) {
          promise.resolve([
            "languageTag": languageTag,
            "ready": true,
          ])
        }
      } catch {
        promise.reject(
          "ERR_ML_KIT_LANGUAGE",
          "Kein ML-Kit-Modell fur \(languageTag) verfugbar.",
          error
        )
      }
    }

    AsyncFunction("recognizeAsync") { (strokes: [[[String: Double]]], languageTag: String, promise: Promise) in
      do {
        let model = try self.getOrCreateModel(languageTag: languageTag)
        self.ensureModelReady(model: model, languageTag: languageTag, promise: promise) {
          do {
            let recognizer = try self.getOrCreateRecognizer(languageTag: languageTag, model: model)
            let ink = self.buildInk(from: strokes)

            recognizer.recognize(ink: ink) { result, error in
              if let error {
                promise.reject(
                  "ERR_ML_KIT_RECOGNIZE",
                  "Die ML-Kit-Erkennung konnte nicht abgeschlossen werden.",
                  error
                )
                return
              }

              promise.resolve(self.buildRecognitionResult(languageTag: languageTag, result: result))
            }
          } catch {
            promise.reject(
              "ERR_ML_KIT_RECOGNIZER",
              "Der ML-Kit-Recognizer konnte nicht initialisiert werden.",
              error
            )
          }
        }
      } catch {
        promise.reject(
          "ERR_ML_KIT_LANGUAGE",
          "Kein ML-Kit-Modell fur \(languageTag) verfugbar.",
          error
        )
      }
    }

    OnDestroy {
      self.recognizerCache.removeAll()
      self.modelCache.removeAll()
    }
  }

  private func getOrCreateModel(languageTag: String) throws -> DigitalInkRecognitionModel {
    if let model = modelCache[languageTag] {
      return model
    }

    guard let identifier = DigitalInkRecognitionModelIdentifier(forLanguageTag: languageTag) else {
      throw NSError(
        domain: "MlKitDigitalInk",
        code: 1,
        userInfo: [NSLocalizedDescriptionKey: "Kein Modell fur \(languageTag)"]
      )
    }

    let model = DigitalInkRecognitionModel(modelIdentifier: identifier)
    modelCache[languageTag] = model
    return model
  }

  private func getOrCreateRecognizer(
    languageTag: String,
    model: DigitalInkRecognitionModel
  ) throws -> DigitalInkRecognizer {
    if let recognizer = recognizerCache[languageTag] {
      return recognizer
    }

    let options = DigitalInkRecognizerOptions(model: model)
    let recognizer = DigitalInkRecognizer.digitalInkRecognizer(options: options)
    recognizerCache[languageTag] = recognizer
    return recognizer
  }

  private func ensureModelReady(
    model: DigitalInkRecognitionModel,
    languageTag: String,
    promise: Promise,
    onReady: @escaping () -> Void
  ) {
    if modelManager.isModelDownloaded(model) {
      onReady()
      return
    }

    var successObserver: NSObjectProtocol?
    var failureObserver: NSObjectProtocol?

    let cleanupObservers = {
      if let successObserver {
        NotificationCenter.default.removeObserver(successObserver)
      }
      if let failureObserver {
        NotificationCenter.default.removeObserver(failureObserver)
      }
    }

    successObserver = NotificationCenter.default.addObserver(
      forName: NSNotification.Name.mlkitModelDownloadDidSucceed,
      object: nil,
      queue: OperationQueue.main
    ) { notification in
      guard
        let downloadedModel = notification.userInfo?[ModelDownloadUserInfoKey.remoteModel.rawValue]
          as? DigitalInkRecognitionModel,
        downloadedModel.modelIdentifier.languageTag == model.modelIdentifier.languageTag
      else {
        return
      }

      cleanupObservers()
      onReady()
    }

    failureObserver = NotificationCenter.default.addObserver(
      forName: NSNotification.Name.mlkitModelDownloadDidFail,
      object: nil,
      queue: OperationQueue.main
    ) { notification in
      guard
        let failedModel = notification.userInfo?[ModelDownloadUserInfoKey.remoteModel.rawValue]
          as? DigitalInkRecognitionModel,
        failedModel.modelIdentifier.languageTag == model.modelIdentifier.languageTag
      else {
        return
      }

      cleanupObservers()
      promise.reject(
        "ERR_ML_KIT_MODEL_DOWNLOAD",
        "Das ML-Kit-Modell fur \(languageTag) konnte nicht geladen werden.",
        nil
      )
    }

    modelManager.download(
      model,
      conditions: ModelDownloadConditions(
        allowsCellularAccess: true,
        allowsBackgroundDownloading: true
      )
    )
  }

  private func buildInk(from strokes: [[[String: Double]]]) -> Ink {
    let inkStrokes = strokes.compactMap { strokePoints -> Stroke? in
      let points = strokePoints.compactMap { point -> StrokePoint? in
        guard let x = point["x"], let y = point["y"] else {
          return nil
        }

        let timestamp = Int(point["t"] ?? Double(Date().timeIntervalSince1970 * 1000))
        return StrokePoint(x: Float(x), y: Float(y), t: timestamp)
      }

      guard !points.isEmpty else {
        return nil
      }

      return Stroke(points: points)
    }

    return Ink(strokes: inkStrokes)
  }

  private func buildRecognitionResult(
    languageTag: String,
    result: DigitalInkRecognitionResult?
  ) -> [String: Any?] {
    let candidates = result?.candidates.map { candidate in
      [
        "score": candidate.score?.doubleValue,
        "text": candidate.text,
      ]
    } ?? []

    return [
      "candidates": candidates,
      "languageTag": languageTag,
      "score": result?.candidates.first?.score?.doubleValue,
      "text": result?.candidates.first?.text,
    ]
  }
}

File: modules/ml-kit-digital-ink/android/src/main/java/expo/modules/mlkitdigitalink/MlKitDigitalInkModule.kt
package expo.modules.mlkitdigitalink

import com.google.mlkit.common.MlKitException
import com.google.mlkit.common.model.DownloadConditions
import com.google.mlkit.common.model.RemoteModelManager
import com.google.mlkit.vision.digitalink.common.RecognitionResult
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognition
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognitionModel
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognitionModelIdentifier
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognizer
import com.google.mlkit.vision.digitalink.recognition.DigitalInkRecognizerOptions
import com.google.mlkit.vision.digitalink.recognition.Ink
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MlKitDigitalInkModule : Module() {
  private val remoteModelManager = RemoteModelManager.getInstance()
  private val modelCache = mutableMapOf<String, DigitalInkRecognitionModel>()
  private val recognizerCache = mutableMapOf<String, DigitalInkRecognizer>()

  override fun definition() = ModuleDefinition {
    Name("MlKitDigitalInk")

    AsyncFunction("isNativeRecognitionAvailableAsync") {
      true
    }

    AsyncFunction("prepareModelAsync") { languageTag: String, promise: Promise ->
      val model = try {
        getOrCreateModel(languageTag)
      } catch (error: Throwable) {
        promise.reject("ERR_ML_KIT_LANGUAGE", error.message, error)
        return@AsyncFunction
      }

      ensureModelReady(languageTag, model, promise) {
        promise.resolve(
          mapOf(
            "languageTag" to languageTag,
            "ready" to true
          )
        )
      }
    }

    AsyncFunction("recognizeAsync") { strokes: List<List<Map<String, Double>>>, languageTag: String, promise: Promise ->
      val model = try {
        getOrCreateModel(languageTag)
      } catch (error: Throwable) {
        promise.reject("ERR_ML_KIT_LANGUAGE", error.message, error)
        return@AsyncFunction
      }

      ensureModelReady(languageTag, model, promise) {
        val recognizer = try {
          getOrCreateRecognizer(languageTag, model)
        } catch (error: Throwable) {
          promise.reject("ERR_ML_KIT_RECOGNIZER", error.message, error)
          return@ensureModelReady
        }

        val ink = buildInk(strokes)

        recognizer
          .recognize(ink)
          .addOnSuccessListener { result ->
            promise.resolve(buildRecognitionResult(languageTag, result))
          }
          .addOnFailureListener { error ->
            promise.reject("ERR_ML_KIT_RECOGNIZE", error.message, error)
          }
      }
    }

    OnDestroy {
      recognizerCache.values.forEach { recognizer ->
        recognizer.close()
      }
      recognizerCache.clear()
      modelCache.clear()
    }
  }

  private fun ensureModelReady(
    languageTag: String,
    model: DigitalInkRecognitionModel,
    promise: Promise,
    onReady: () -> Unit
  ) {
    remoteModelManager
      .isModelDownloaded(model)
      .addOnSuccessListener { isDownloaded ->
        if (isDownloaded) {
          onReady()
          return@addOnSuccessListener
        }

        remoteModelManager
          .download(model, DownloadConditions.Builder().build())
          .addOnSuccessListener {
            onReady()
          }
          .addOnFailureListener { error ->
            promise.reject(
              "ERR_ML_KIT_MODEL_DOWNLOAD",
              "Das ML-Kit-Modell fur $languageTag konnte nicht geladen werden.",
              error
            )
          }
      }
      .addOnFailureListener { error ->
        promise.reject(
          "ERR_ML_KIT_MODEL_STATUS",
          "Der ML-Kit-Modellstatus fur $languageTag konnte nicht gepruft werden.",
          error
        )
      }
  }

  private fun getOrCreateModel(languageTag: String): DigitalInkRecognitionModel {
    modelCache[languageTag]?.let { return it }

    val modelIdentifier = try {
      DigitalInkRecognitionModelIdentifier.fromLanguageTag(languageTag)
    } catch (error: MlKitException) {
      throw IllegalArgumentException("Kein ML-Kit-Modell fur Sprache $languageTag.", error)
    } ?: throw IllegalArgumentException("Kein ML-Kit-Modell fur Sprache $languageTag.")

    return DigitalInkRecognitionModel
      .builder(modelIdentifier)
      .build()
      .also { model ->
        modelCache[languageTag] = model
      }
  }

  private fun getOrCreateRecognizer(
    languageTag: String,
    model: DigitalInkRecognitionModel
  ): DigitalInkRecognizer {
    recognizerCache[languageTag]?.let { return it }

    return DigitalInkRecognition
      .getClient(DigitalInkRecognizerOptions.builder(model).build())
      .also { recognizer ->
        recognizerCache[languageTag] = recognizer
      }
  }

  private fun buildInk(strokes: List<List<Map<String, Double>>>): Ink {
    val inkBuilder = Ink.builder()

    strokes.forEach { strokePoints ->
      if (strokePoints.isEmpty()) {
        return@forEach
      }

      val strokeBuilder = Ink.Stroke.builder()

      strokePoints.forEach { point ->
        val x = point["x"]?.toFloat() ?: return@forEach
        val y = point["y"]?.toFloat() ?: return@forEach
        val t = point["t"]?.toLong() ?: System.currentTimeMillis()
        strokeBuilder.addPoint(Ink.Point.create(x, y, t))
      }

      inkBuilder.addStroke(strokeBuilder.build())
    }

    return inkBuilder.build()
  }

  private fun buildRecognitionResult(
    languageTag: String,
    result: RecognitionResult
  ): Map<String, Any?> {
    val candidates = result.candidates.map { candidate ->
      mapOf(
        "score" to null,
        "text" to candidate.text
      )
    }
    val topText = result.candidates.firstOrNull()?.text

    return mapOf(
      "candidates" to candidates,
      "languageTag" to languageTag,
      "score" to null,
      "text" to topText
    )
  }
}
